Searchable attachments

wielinde opened the pull request.

wielinde committed:
WIP: attachment search

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/ebe1c0515215a51833026a0aeb2a34203a4d33c9) (deployed commit: 63f3c4cb02e0727985f78ff5bee5c175c47ebe17).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/50e8cfa39b5278d265858093f76eee4d7ca155c4) (deployed commit: 0026efca90a351c741ac86aa142abb3d96c749f2).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/f9d6ea5cf3c38ffe0db6671560edb3015ce9e0ff) (deployed commit: 4fe7c6363f9ef53f955f4a76b6ba2139fc6a880a).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/11200bce4f9609a236df05d663d2f548156e1bf7) (deployed commit: db8fdfb6126b1f1ae5770dafaf6fc8609122f9a9).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/0b20bb73999a64fe64ae1d4a691767298e72f8e0) (deployed commit: db8fdfb6126b1f1ae5770dafaf6fc8609122f9a9).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/6d4389eb9f00cb88f625b1c3c49475e978ed9e21) (deployed commit: b1510200157b923571dc55bf66b128df25f255e1).

wielinde committed:
adding more fixture files

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/d7add0f014766339b4b6a8a6ba7c3de120d64619) (deployed commit: cd8008c45a231ebde66d6740f7cab2a05e4b5999).

wielinde committed:
remove fixture file for text extraction (and move them to separate gem)

wielinde committed:
Adopt to new text-extractor gem

wielinde committed:
Add first LIKE based WP filter for Attachment content and file name

wielinde committed:
Allow filter WPs based on the content or file name of their attachments

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/da009f019230a8803383f7b41994f5c30ff10cf3) (deployed commit: 4f2b6d2e47a37865787dceff0330f71e81c6a5ec).

wielinde committed:
Move installation text extraction packages to Dockerfile

wielinde committed:
Install postgresql-client, mysql-client and sqlite to allow for dumping schemas.

wielinde committed:
Fix bug missing includes method for shared custom field filters

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/441ecc60bd89f8db0b42863df807e10d6350de9c) (deployed commit: 13764c8aeae31359e233b1d61c11333878d6f5d5).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/ad71f1b9369f108bbcc26656dd7c243f4adf638a) (deployed commit: 13764c8aeae31359e233b1d61c11333878d6f5d5).

wielinde committed:
Trying to have the same postgresql-client version as server

wielinde committed:
Add spec for ExtractFulltextJob

wielinde committed:
Adding pretty generic specs for attachment filters.

wielinde committed:
trying a fresher docker base image that has postgresql-client 9.6 by default

wielinde committed:
Remove typo RUN

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/d914a7727aa8a3f239277c2845922853ea8d243b) (deployed commit: c7af5d1b8ef991cae33a885854881ae2f227de99).

wielinde committed:
- Remove configuration option for attachment search
- Remove attachment search from global search as it is currently incompatible
- Add rake task for indexing attachments that have not been indexed yet
- Make the attachments filters a EE feature.

ulferts commented on Gemfile:
```diff
@@ -162,6 +162,7 @@ gem 'carrierwave', '~> 1.2.2'
 gem 'aws-sdk', '~> 2.10.1'
 
 gem 'openproject-token', '~> 1.0.1'
+gem 'text-extractor', git: 'https://github.com/planio-gmbh/text-extractor'
```
Especially when using a git repos, we should fix the version by providing a `ref:`

wielinde replied:
Yes, this is to keep things easy until we release the gem on rubygems. I do not want to merge into dev until we release the gem.

ulferts commented on app/models/attachment.rb:
```diff
@@ -178,4 +180,19 @@ def self.content_type_for(file_path, fallback = OpenProject::ContentTypeDetector
+
+  def extract_fulltext
+    job = ExtractFulltextJob.new(id)
+    Delayed::Job::enqueue job
```
This is basically the same as in the [class method](https://github.com/opf/openproject/pull/6038/files#diff-7b0581da1680146a96496be759fc3240R194) but uses an interface not agnostic to Delayed::Job.

wielinde replied:
It is not. The class method does not queue the jobs to a later execution but executes them directly as a batch.

In other places where we enqueue jobs we use `Delyaed::Job::enqueue`, too. An agnostic way would be to call `perform_later` but that class method is not available for some weird reason. Grrr.

ulferts commented on app/models/queries/work_packages.rb:
```diff
@@ -34,6 +34,10 @@ module Queries::WorkPackages
   register.filter Query, filters_module::AssigneeOrGroupFilter
+  if EnterpriseToken.allows_to?(:attachment_filters)
+    register.filter Query, filters_module::AttachmentContentFilter
+    register.filter Query, filters_module::AttachmentFileNameFilter
```
This is the reason why the migrations currently fail, I believe. The filters have an `#available?` method where this functionality would be situated better. Doing this would also remove the necessity for a restart after the user enters an enterprise token in a production setting.

ulferts commented on app/models/queries/work_packages/filter/attachment_base_filter.rb:
```diff
@@ -0,0 +1,73 @@
+  private
+
+  def tokenize
+    terms = I18n.transliterate(values.first.downcase).split /\s/
```
Haven't tried it out yet but I think transliterating here is not a good idea. The database information as well as the input should all be in utf8 so they should be compatible. On the other hand:

```
I18n.transliterate('Gänseblümchen')
=> "Ganseblumchen"
```

which unless the whole of the ts_vector is also transliterated, should result in not finding `Gänseblümchen`. And if the vector is transliterated. It will result in false matches and faulty text to be displayed if we ever implement showing text extracts.

ulferts replied:
If we find out that this is a good idea after all `OpenProject::FullTextSearch.normalize_text` should be used here.

wielinde replied:
Good catch. I should use the methods from `OpenProject::FullTextSearch`.

The transliteration is necessary, as we want that "Ganseblumchen" finds "Gänseblümchen". One rightfully argue that this is something the DB should take care of. Unfortunately we need the Postgres Contrib module UNACCENT for this. If we can do it in our code, we do not have the dependency to the contib package.

wielinde replied:
So both, the filter value and the extracted text need to pass the normalization.

ulferts commented on app/models/queries/work_packages/filter/attachment_content_filter.rb:
```diff
@@ -0,0 +1,50 @@
+  end
+
+  def human_name
+    I18n.t('label_attachment_content')
```
It probably boils down to personell preference but I strongly dislike the use of `label_...` as it suggests this to be used as some kind of label only (strictly speaking, only to be used in a `label_tag`) but that is by far not the case in OpenProject. 
I believe it is better to describe what is in the string. And it this case, it is just the content of the attachment, so 
`Attachment.human_attribute_name('content')` makes most sense to me.
This has the added benefit, that it will be used in a couple of rails form helpers.

ulferts commented on app/models/queries/work_packages/filter/attachment_file_name_filter.rb:
```diff
@@ -0,0 +1,50 @@
+  end
+
+  def human_name
+    I18n.t('label_attachment_file_name')
```
please see above.

ulferts commented on app/models/work_package.rb:
```diff
@@ -123,8 +123,8 @@ class WorkPackage < ActiveRecord::Base
-                     include: %i(project journals),
-                     references: %i(projects journals),
+                     include: %i(project journals).push(:attachments) ,
+                     references: %i(projects journals).push(:attachments),
```
Why is this here? It looks as if it belongs to a different PR.

wielinde replied:
👍

ulferts commented on app/workers/extract_fulltext_job.rb:
```diff
@@ -0,0 +1,80 @@
+require 'text_extractor'
+
+class ExtractFulltextJob < ApplicationJob
+  # queue_as :text_extraction
```
please delete?

wielinde replied:
👍

ulferts commented on app/controllers/admin_controller.rb:
```diff
@@ -76,7 +76,12 @@ def info
+      # [:text_unrtf_available,          OpenProject::TextExtractor::RtfHandler.available?],
+      # [:text_catdoc_available,         OpenProject::TextExtractor::DocHandler.available?],
+      # [:text_xls2csv_available,        OpenProject::TextExtractor::XlsHandler.available?],
+      # [:text_catppt_available,         OpenProject::TextExtractor::PptHandler.available?]
```
commented out code

wielinde replied:
👍

ulferts commented on app/workers/extract_fulltext_job.rb:
```diff
@@ -0,0 +1,80 @@
+    text = TextExtractor::Resolver.new(attachment.diskfile, attachment.content_type).text if attachment.readable?
+
+    if OpenProject::Database.allows_tsv?
+      attachment_filename = attachment.file.file.filename
```
You lost me after the first `file` :smile:
[Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter)?

wielinde replied:
hehe, the issue is here that `attachment.file` is not the file but the Carrierwave uploader. I changed the code to improve the readability. While not changing the structure.

ulferts commented on app/workers/extract_fulltext_job.rb:
```diff
@@ -0,0 +1,80 @@
+  def perform
+    return unless attachment = find_attachment(@attachment_id)
+
+    text = TextExtractor::Resolver.new(attachment.diskfile, attachment.content_type).text if attachment.readable?
```
Have you tried this while having S3 configured as your attachment backend? The attachments [will soon be deleted after they have been updated to S3](https://github.com/opf/openproject/commit/fb052d21258aa66384291ca32e82e0dd241df157#diff-e9137f5f1f8f8a77cd7eead296d83830) so your job will very likely not find the attachment locally any more.

wielinde replied:
No, I haven't and I need help here.

ulferts commented on app/workers/extract_fulltext_job.rb:
```diff
@@ -0,0 +1,80 @@
+  private
+
+  def update_with_tsv(id, filename, text)
+    language = OpenProject::Configuration.main_content_language
```
This is prone to break in a SaaS setting or more generally speaking in a multi language environment. Do we have additional requirements on the db depending on what languages we support?

ulferts commented on config/environments/production.rb:
```diff
@@ -75,7 +75,7 @@
 
   # Set to :debug to see everything in the log.
-  config.log_level = :warn
+  # config.log_level = :warn
```
This change is probably unnecessary

ulferts replied:
should still be undone before merging

ulferts commented on config/application.rb:
```diff
@@ -132,7 +132,7 @@ class Application < Rails::Application
     # This is necessary if your schema can't be completely dumped by the schema dumper,
     # like if you have constraints or database-specific column types
-    # config.active_record.schema_format = :sql
+    config.active_record.schema_format = :sql
```
What is the rationell for this?

wielinde replied:
Apparently ActiveRecord does not know how to dump TSVector column definition to `schema.rb`.

ulferts commented on config/initializers/delayed_job_config.rb:
```diff
@@ -31,4 +31,5 @@
 # By default bypass worker queue and execute asynchronous tasks at once
-Delayed::Worker.delay_jobs = !Rails.env.test?
+# Delayed::Worker.delay_jobs = !Rails.env.test?
+Delayed::Worker.delay_jobs = false
```
Probably a leftover from development

ulferts replied:
should still be undone before merging

ulferts commented on config/locales/en.yml:
```diff
@@ -2112,6 +2114,12 @@ en:
+  text_unrtf_available: "Unrtf available (optional)"
+  text_catdoc_available: "Catdoc available (optional)"
+  text_xls2csv_available: "Xls2csv available (optional)"
+  text_catppt_available: "Catppt available (optional)"
```
Personell preference again but I prefer scopes e.g.

```
extraction:
 available:
  pdftotext: ...
  ...
```
As it tends to keep translations belonging together in one place.

ulferts commented on db/migrate/20180122135443_add_tsv_columns_to_attachments.rb:
```diff
@@ -0,0 +1,24 @@
+      remove_column :attachments, :fulltext_tsv
+      remove_column :attachments, :file_tsv
+    end
+  end
```
You should be able to write this with a change

```
def change
 if OpenProject::Database.allows_tsv?
  ...
 else
  reversible do |dir|
    dir.up do
     warn "Your installation does not support full-text search features. Better use PostgreSQL in version 9.6 or higher."
    end
   end

end
```

wielinde replied:
Nice. Man lernt nicht aus.

ulferts commented on docs/operations/upgrading/manual/upgrading.md:
````diff
@@ -226,6 +226,15 @@ The actual upgrade commands are as follows:
+run:
+
+```bash
+[openproject@debian]# RAILS_ENV="production" rake attachments:extract_fulltext
````
This should enable us to trigger this as a series of background jobs as opposed to having to wait for the migrations to finish. LGTM :+1:

ulferts commented on app/workers/extract_fulltext_job.rb:
```diff
@@ -0,0 +1,80 @@
+            "fulltext_tsv" = to_tsvector('%s', '%s'),
+            "file_tsv" = to_tsvector('%s', '%s')
+          WHERE ID = %s;
+    SQL
```
```
Attachment
 .where(id: id)
 .update_all(['fulltext = ?, fulltext_tsv = to_tsvector(?, ?), file_tsv = to_tsvector(?, ?), 
             text, 
             language, 
             OpenProject::FullTextSearch.normalize_text(text), 
             language, 
             OpenProject::FullTextSearch.normalize_text(filename)])
``` 
should do the same while sanitizing at the same time.

ulferts replied:
As written above, I believe transforming to ascii to be faulty and unnecessary. In addition, I could imagine it being slow when a file having 1 million chars is processed like this.

ulferts reviewed (changes requested):
I haven't yet tried out the code but I thought it worthwhile to give already provide an initial review so that you are not stuck waiting for my input.

IMO it would additionally make sense to have a feature spec employing the filters.

wielinde committed:
Use filter instance method `available?` for checking Enterprise Token

wielinde committed:
Add EnterpriseToken logic to attachment filter specs

ulferts commented on app/models/queries/work_packages/filter/attachment_base_filter.rb:
```diff
@@ -0,0 +1,78 @@
+        :sanitize_sql_array,
+        [where_condition, search_column, language, tokenize]
+      )
+    else
```
![image](https://user-images.githubusercontent.com/617519/35514162-980a9376-0505-11e8-9658-50ed4ae5c36d.png)

wielinde replied:
Ugly

wielinde replied:
In my opinion this is not a security issue. However, it is bad that invalid Query get through down to the DB level. Now I clean/filter the input for trouble making characters first.

wielinde committed:
Fix the commit of text-extractor gem in Gefiel

wielinde committed:
Fix Docker entrypoint to cope with structure.sql requirement (instead of schema.rb)

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/9c1663a92d5b914a8f0a2a88ae0da4011be2a514) (deployed commit: 034ab92bb20efbbb7ffaf81652ba2f82b9ab5d26).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/822b9f4ad38b61859de7a27d4dae95e2b5c35666) (deployed commit: 034ab92bb20efbbb7ffaf81652ba2f82b9ab5d26).

wielinde committed:
Show availability of third party packages for text extraction

wielinde committed:
Cleanup

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/acc3bff0f9610019e04ed4d213988c1149462f4a) (deployed commit: 034ab92bb20efbbb7ffaf81652ba2f82b9ab5d26).

wielinde committed:
normalize file names differently

wielinde committed:
Improve code readability

wielinde committed:
Add documentation for configuration options

wielinde committed:
code cleanup

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/82a5d4a9ae4136407053456fc0757d0451215199) (deployed commit: 01712b8d4e267fd2f1d25372f4755b7be796c33b).

wielinde committed:
remove structure.sql from version control

wielinde committed:
ignore structure.sql in version control

wielinde committed:
Fix for not crashing DB query when search term is "')"

wielinde committed:
restructure ExtractFulltextJob for better readability

wielinde committed:
Provide second rake task that updates text extractions even for those that have already been extracted.

wielinde committed:
to_words needs to be class method.

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/8f0e02d0c1e877971c6abf458ccec98b23e7bbfc) (deployed commit: 35b2cf3b86cc6ae52dc8375a9f934b4d489f0fd7).

wielinde committed:
Add feature spec for attachment filters

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/fff459fb941a1fb01d1e45173540ec8dfd626290) (deployed commit: 290642d76cd1a69839e80137f3621ce4f28648f7).

wielinde committed:
move rake tasks to attachments name space

wielinde committed:
Satisfy rubucop

wielinde committed:
cleanup i18n

wielinde committed:
DRY migration

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/9b31d0c8f445bcc11d77c3a5e833a3ee0ffd03c8) (deployed commit: ac8811f308e6fcc4fa079f21d50c83cc74ac4fb2).

wielinde committed:
Satisfy or calm rubocop

wielinde committed:
Add upsale message

wielinde committed:
Do not allow attachment filters unless TSVector is available. No MySQL/MariaDB support.

wielinde commented:
@ulferts thank you for your review. Could you please checkout the latest commits?

BTW: I changed the feature a bit:
- No non TSVector support anymore.
- Added EE upsale banner
- File names get sliced into words, so a filter on "foo" would still find "foo-bar.pdf"

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/60ff8a939e8a28e50c2c836bd934ada460cae9df) (deployed commit: 01298c883691372e0cd55da8ac751fef37444d77).

wielinde committed:
Adopt code for API changes in text-extractor gem
- remove obsolete require statements
- handle exceptions
- adopt Specs

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/8c8f043dc3cda9f6a27a6cf8aacfb868812c8f68) (deployed commit: 4374c742f38ecc858763fbdbc80455668760b9e3).

wielinde committed:
- Use initializer to load text-extractor config.
- Support OCR for image attachments.

wielinde committed:
Bump text-extractor gem

wielinde committed:
Fix update_form_api_spec as `allows_to? :attachment_filter` gets called

wielinde committed:
Fix: allow spec to call EnterpriseToken for checking :attachment_filters

wielinde committed:
Fix: These tests should not do anything when the DB does not support TSVector

wielinde committed:
bump text-extractor gem reference

wielinde committed:
Add guard condition for TSVector support in extraction job. Remove non TSV support.

wielinde committed:
Fix spec. Check that attachment filters are not selectable when DB does not support TSVector.

wielinde committed:
Fix spec: only run those specs if DB supports TSVector.

wielinde committed:
Use rubygems.org version of text-extractor.

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/875ea55bfcd9d2a3b0316b3db8e2ca4c491fa718) (deployed commit: 68d4f36a15d99976f0ffff53bbfc5db0451bc738).

ulferts commented on app/controllers/admin_controller.rb:
```diff
@@ -76,7 +76,13 @@ def info
+      [:'extraction.available.catdoc',    TextExtractor::DocHandler.available?],
+      [:'extraction.available.xls2csv',   TextExtractor::XlsHandler.available?],
+      [:'extraction.available.catppt',    TextExtractor::PptHandler.available?],
+      [:'extraction.available.tesseract', TextExtractor::ImageHandler.available?]
```
As we disable the feature completely when 
* running on MySQL 
* not having an enterprise token

it would probably make sense to also note it down here.
We might want to disable the other checks altogether if the precondition (postgresql and enterprise token) is not met.

wielinde replied:
Yes, agreed.

One detail though: If `allows_tsv?` returns `true` we currently do extract attachments even without having a token present. That way, if someone adds a token the feature is up and running straight away without the need of bothering a system adminstrator to start a rake task.

ulferts replied:
👍

ulferts commented on app/models/query/results.rb:
```diff
@@ -209,14 +209,18 @@ def transform_custom_field_keys(custom_field, groups)
   # * sorting
   # * grouping
-  def include_columns
+  def include_columns # rubocop:disable Metrics/AbcSize
```
disabling rubocop should not be done. Please restructure the code to get below 15.0

wielinde replied:
Well, I think it is nicely readable code. However, I changed it to something worse that satisfies the rubocop criteria ;-)

ulferts commented on lib/open_project/database.rb:
```diff
@@ -144,5 +144,12 @@ def self.numeric_version
+    # search.
+    def self.allows_tsv?
+      OpenProject::Database.name == :postgresql &&
+        Gem::Version.new(OpenProject::Database.version) >= Gem::Version.new('9.6')
```
Do we actually need 9.6 or does 9.5 suffice? We currently have 9.5 noted down in the system requirements and it would be nice (but not necessary) if the feature would work with 9.5 as well.

wielinde replied:
True, 9.5 or even older should work. I got confused with the phrase search that was introduced in 9.6 and our internal buzz about 9.6 (docker update etc) did not help.

ulferts commented on spec/models/attachment_spec.rb:
```diff
@@ -73,6 +73,17 @@
+    #               :content_type => 'text/plain')
+    #   a.reload
+    #   assert a.fulltext.include?("this is a text file for upload tests\r\nwith multiple lines")
+    # end
```
Commented out test. Either remove if enough unit tests exist or turn into spec.

ulferts commented on app/workers/extract_fulltext_job.rb:
```diff
@@ -0,0 +1,80 @@
+    @text = nil
+    @file = nil
+    @filename = nil
+    @language = OpenProject::Configuration.main_content_language
```
I can unfortunately no longer recall what we decided regarding having the language as a configuration as opposed to having it as a setting. The later would allow the user to alter it via the UI but might require us to rerun the ts_vector rake task upon changes to it. For SaaS however, this will probably be necessary as english might not always be the correct setting. I therefore think it is "alternativlos"
 
![image](https://user-images.githubusercontent.com/617519/36250156-46a2cb58-123d-11e8-82c6-a85003cd6d7c.png)

wielinde replied:
In general terms you are right. However, I did not want to bother too much about that issue at the moment. It would get complicated pretty quickly. And I am hopeful that the filter will work even with the wrong language selected. I just indexed Spanish documents with an English configuration and as long as the filter query uses the same language thinks work. Maybe I don't see the issues bubble up yet but I would not care too much at the moment. When we implement not only filters but also fulltext search maybe this should receive more attention.

Further, I imagine that reindexing the extracted fulltext can take a while. We do not have any means to communicate these long running jobs to an admin user, do we?

ulferts replied:
We don't communicate it and that is already a problem. E.g. the rake task in this PR will suffer from the same problem already as we provide no feedback there as well. E-mail sending would be another example where it would make sense to provide feedback. But as we already have that problem I wouldn't, having the same problem again for reindexing does not sound too bad to me.

But maybe it is ok to roll with what we have (others have done so too: https://github.com/zotonic/zotonic/commit/c8aad274489029dc5d8928c6781034253ca9d00e) and see how far we can get. 

But let's keep the topic in the back of our heads once additional features like trigrams are added which will rely on the language more heavily.

wielinde replied:
@ulferts I am not sure if understand everything.

Yes, the rake tasks could be more informative. But I did not understand the rest of your explanation. Let's quickly catch up on this on Monday.

ulferts reviewed (changes requested):
Still works nicely!

Appart from my comments, there are two code climate complaints.

wielinde committed:
Gem name changed from ‚text_extractor‘ to ‚plaintext‘

wielinde committed:
rubocop stuff and cleaning up code

wielinde committed:
Enforce lower PostgreSQL version for TSVector features, 9.5 instead of 9.6

wielinde committed:
Adopt system checks into two cases, with tsvector support and without.

wielinde committed:
more rubocop stuff

ulferts commented on lib/open_project/full_text_search.rb:
```diff
@@ -0,0 +1,47 @@
+  # This module provides utility methods to work with PostgreSQL's full-text capabilities (TSVECTOR)
+  module FullTextSearch
+    def self.normalize_text(text)
+      I18n.transliterate(text.to_s.downcase)
```
I just stumbled over https://apidock.com/rails/ActiveSupport/Inflector/transliterate. That means that transliteration can (does not have to be) language dependent. Maybe I am paranoid here but ensuring that transliteration always happens in the same language, probably `OpenProject::Configuration.main_content_language`, would be safer:
```
I18n.with_locale OpenProject::Configuration.main_content_language do
 I18n.transliterate ...
end

wielinde replied:
Oh wow, didn't know that. Good find. However, I simplified this to always use the "en" locale. It does not really matter which transliteration is used as long as it is consistent. I checked German Umlauts and sharp S and both were transliterated the same for "en" and "de".

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/47d0ea41683e91b650277fbc454510d0b4bcdca2) (deployed commit: 1e468b3cbf94eb73cd15f2460cd68f1966b53cff).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/f957215cae44f4ae221c6378f9dc2c8fd8b14cc7) (deployed commit: 1e468b3cbf94eb73cd15f2460cd68f1966b53cff).

wielinde committed:
Add tesseract-ocr to Dockerfile

wielinde committed:
Ensure that always the identical transliteration is used no matter what I18n locale the current user has set.

wielinde committed:
Debugging pull preview

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/4d32497cee03fe1297c67edc4d85c976b665a955) (deployed commit: 14e88ad4ea4c6466832066278077a99bfa9f95ef).

wielinde committed:
Rubocop satisfaction

ulferts commented on config/plaintext.yml:
```diff
@@ -0,0 +1,38 @@
+# Text extraction helper programs.
```
As the plaintext.yml was added to gitignore, it should probably not have been checked in.

wielinde replied:
Good catch. Thanks!

wielinde committed:
remove plaintext.yml from version control that was mistankenly checked in

ulferts committed:
propose pgsql 9.5 instead of 9.6
